### PR TITLE
PL: Ingest misc surveys

### PR DESCRIPTION
**Pull request by breville**
This adds a model for miscellaneous surveys: `Pd::MiscSurvey`.  

It also starts ingesting six of these surveys from JotForm.  These initial surveys were filled out by end-users directly in JotForm, and while most of them have user IDs, not all do.  They don't have environments (test/production/etc.) either.

The idea of this model is that we can increasingly make surveys "self-serve" by internal team members when they need to send out a survey, and want it ingested into our database for post-processing, but there isn't additional code in our system that will depend on their submitted contents.

One improvement for readability is that we store answers using the `name` of the question as the key, rather than `id`.  That means where we have previously stored survey submissions like this:

```
{4=>"Alfred",
 5=>"1234567",
 6=>"CS Discoveries",
 8=>"2018 CS Discoveries Facilitator",
 9=>"Specific Partner Name",
 10=>"1",
 12=>"Yes",
 13=>"No",
 16=>
  "Yes, I taught this course as a classroom teacher in the 2018-19 school year.",
```

They are now stored like this:

```
{"userName"=>"Alfred",
 "userId"=>"1234567",
 "program"=>"CS Discoveries",
 "track"=>"2018 CS Discoveries Facilitator",
 "regionalPartnerName"=>"Specific Partner Name",
 "makeup"=>"1",
 "ayw_work"=>"Yes",
 "sw_work"=>"No",
 "currentlyTeachingCourse"=>
  "Yes, I taught this course as a classroom teacher in the 2018-19 school year.",
```

I really dislike that we don't have a similar concept of a human-readable key for multiple-choice answers, and are instead storing raw strings for every response, making them duplicated across answers to the same survey.  But JotForm doesn't seem to support this concept, and auto-generating human-readable keys seems challenging to get right.

### CODE REVIEW

File: dashboard/app/models/concerns/pd/jot_form_backed_form.rb [hunk 1]
```diff
@@ -169,10 +177,29 @@ def _sync_from_jotform(form_ids)
         imported
       end
 
-      def process_submission(submission)
+      # Process a submission and saves the model.
```
Comment (clareconstantine) [inline dashboard/app/models/concerns/pd/jot_form_backed_form.rb:180]: yay for adding comments!

File: dashboard/db/migrate/20190521215856_create_pd_misc_surveys.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+class CreatePdMiscSurveys < ActiveRecord::Migration[5.0]
```
Comment (clareconstantine) [inline dashboard/db/migrate/20190521215856_create_pd_misc_surveys.rb]: can we separate the migration into its own PR?

**Review by clareconstantine — APPROVED** [review]
To clarify my understanding, is this our new generic survey model for all future surveys? or just for surveys that are not embedded in our site?

**Comment by breville** [comment]
While its first usage here is for survey that are not embedded on our site, I'm now working on a survey that is embedded on our site that will store to this model too.  So I see this model working for both cases, but only when we don't want to save to a more specific model because we need significant business logic associated with the data, for example.

I will separate the migration into its own PR for sure; just wanted to initially present it as part of a whole.

**Comment by islemaster** [comment]
So with this new system, as a survey author, what steps do I take to build and launch a survey? When do I need to get engineering involved?

File: dashboard/app/models/concerns/pd/jot_form_backed_form.rb [hunk 3]
```diff
@@ -107,6 +107,14 @@ def _sync_from_jotform(form_ids)
 
         form_ids.each do |form_id|
           questions = get_questions(form_id, force_sync: true)
+          questions_details = use_names_for_question_ids? ? JSON.parse(questions.questions) : nil
+          if questions_details
+            # Make sure that there is a unique, non-nil name for each question.
+            if questions_details.pluck("name").compact.uniq.size != questions_details.size
+              raise "Not all questions for form #{form_id} have unique names."
```
Comment (islemaster) [inline dashboard/app/models/concerns/pd/jot_form_backed_form.rb:114]: I'm interested in what we can do to surface feedback like this to survey authors as early as possible.

One idea: Provide a test route (e.g. `studio.code.org/validate_jotform`) that, given a jotform id, will "lint" the form to ensure it only uses the subset of features we support, and returns red/green so form authors can check "is my form ok?" without asking an engineer or waiting for us to surface a Honeybadger error.

**Comment by breville** [comment]
This is still in the phase of laying foundations, but in the parallel work I'm doing, it's as simple as sharing a URL like this: https://studio.code.org/pd/misc_survey/[form_id] if you want to host the survey on our site.  And whether we host or not, engineering just need to add the survey ID to the list in `Pd::MiscSurvey` for subsequent ingestion.

There seems a natural roadmap that follows: a more centralised list of survey IDs we care about, specifying properties in that list for each survey regarding things such as whether they require a user to be signed-in, whether a user should be a teacher, etc. but I like the idea of adding them as requirements arise, and keeping it minimal ahead of then.

**Comment by islemaster** [comment]
Awesome! I could even see that list of surveys living in our database, maybe seeded from a csv file like videos to maintain the list across environments.  Then authors could enable/disable/customize survey metadata as well through levelbuilder.

File: dashboard/app/models/concerns/pd/jot_form_backed_form.rb [hunk 4]
```diff
@@ -169,10 +177,29 @@ def _sync_from_jotform(form_ids)
         imported
       end
 
-      def process_submission(submission)
+      # Process a submission and saves the model.
+      # @param [submission] (optional) The submission received from JotForm.
```
Comment (islemaster) [inline dashboard/app/models/concerns/pd/jot_form_backed_form.rb:181]: Is `submission` actually optional?

File: dashboard/app/models/concerns/pd/jot_form_backed_form.rb [hunk 5]
```diff
@@ -169,10 +177,29 @@ def _sync_from_jotform(form_ids)
...
+      def process_submission(submission, questions_details=nil)
         # There should be no duplicates, but just in case handle them gracefully as an upsert.
         find_or_initialize_by(submission.slice(:form_id, :submission_id)).tap do |model|
-          model.answers = submission[:answers].to_json
+          dest_answers = {}
+
+          source_answers = submission[:answers]
+          source_answers.each do |key, value|
+            # We have an answer ID, but can we get a name instead of the ID?
+            question_detail = questions_details && questions_details.find {|q| q["id"] == key}
```
Comment (islemaster) [inline dashboard/app/models/concerns/pd/jot_form_backed_form.rb:193]:
```suggestion
            question_detail = questions_details&.find {|q| q["id"] == key}
```

**Review by islemaster — APPROVED** [review]
Woohoo!

**Comment by hacodeorg** [comment]
Talked to Brendan offline, summarize it here. This model fills one of the gaps in our survey ingesting approach. 

We still need specialized or additional models to support submitting a form multiple times for different contexts (e.g. same user different workshops, or different days in the same workshop) and/or denying submitting a form twice in the same context.

I have 2 main concerns with the current way of ingesting data (not specific to this model).
1. The separation of questions and answers. We save questions and answers in 2 different models, and only save the latest version of questions. Answers will become invalid when questions change or are removed. We should come up with a model that makes submissions self-contained.

2. The mixing of downloading, validating and processing submissions. We are doing several jobs in 1 transaction; if any of them fails the whole thing is thrown away and we retry next time.

File: dashboard/app/models/concerns/pd/jot_form_backed_form.rb (hunk 3 above)
Comment (hacodeorg) [reply dashboard/app/models/concerns/pd/jot_form_backed_form.rb:114]: +1 to prevention. Since we only detect this after there are submissions to the forms, we will have to update those submissions in JotForm if we want to ingest them.

**Comment by hacodeorg** [comment]
Thinking more about this, the main reason why we need the specialized models are because there are context-specific keys like `day` and `facilitator_id`. What if we add an optional `custom_key` to the MiscSurvey model so it can cover all these cases? This will solve both issues, allowing multiple submissions for different contexts and preventing double submissions for the same context.

```
#  id            :integer          not null, primary key
#  form_id       :integer          not null
#  submission_id :integer          not null
#  answers       :text(65535)
#  user_id       :integer
#  custom_key: text(65535)
```

P/S: using `day` (a single number 0->5+) to signal pre-workshop, daily and post-workshop surveys is problematic. For example, in CSF201 (Deep dive) day1 means post-workshop, what if CSF201 becomes a 2-day thing? For CSF101 (Intro) right now have 1 survey, so it will be day0, what if it needs pre-workshop survey also?

**Comment by hacodeorg** [comment]
`custom_key` can be a composite keys such as `{"type"=>"day1", "facilitator_id"=>"3000"}`. We can then `sort` the hash and save it as string of array `[["facilitator_id", "3000"], ["type", "day1"]]`, pass it along to JotForm and have it sent back to us to create placeholder.

File: dashboard/app/models/concerns/pd/jot_form_backed_form.rb (hunk 4 above)
Comment (breville) [reply dashboard/app/models/concerns/pd/jot_form_backed_form.rb:181]: Will address in https://github.com/code-dot-org/code-dot-org/pull/28850.

File: dashboard/app/models/concerns/pd/jot_form_backed_form.rb (hunk 5 above)
Comment (breville) [reply dashboard/app/models/concerns/pd/jot_form_backed_form.rb:193]: Will address in https://github.com/code-dot-org/code-dot-org/pull/28850.
